src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs

```diff
@@ -0,0 +1,155 @@
+            var completion = new TaskCompletionSource<TResponseType>();
+
+            var item = new QueueItem(mutatesSolutionState, clientCapabilities, clientName,
+                callback: async (context, cancellationToken) =>
+                {
+                    // Check if cancellation was requested while this was waiting in the queue
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        completion.SetCanceled();
+
+                        // Tell the queue to ignore any mutations from this request, not that we've given it a chance
+                        // to make any
+                        return false;
+                    }
+
+                    try
+                    {
+                        var result = await handler.HandleRequestAsync(request, context, cancellationToken).ConfigureAwait(false);
+                        completion.SetResult(result);
+                        // Tell the queue that this was successful so that mutations (if any) can be applied
+                        return true;
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        completion.SetCanceled();
+                    }
+                    catch (Exception exception)
+                    {
+                        // Pass the exception to the task completion source, so the caller of the ExecuteAsync method can observe but
+                        // don't let it escape from this callback, so it doesn't affect the queue processing.
+                        completion.SetException(exception);
+                    }
+
+                    // Tell the queue to ignore any mutations from this request
+                    return false;
+                }, requestCancellationToken);
+            _queue.Enqueue(item);
+
+            return completion.Task;
+        }
+
+        private async Task ProcessQueueAsync()
+        {
+            // Keep track of solution state modifications made by LSP requests
+            Solution? lastMutatedSolution = null;
+            var queueToken = _cancelSource.Token;
+
+            while (!queueToken.IsCancellationRequested)
+            {
+                var work = await _queue.DequeueAsync().ConfigureAwait(false);
+
+                // Create a linked cancellation token to cancel any requests in progress when this shuts down
+                var cancellationToken = CancellationTokenSource.CreateLinkedTokenSource(queueToken, work.CancellationToken).Token;
+
+                // The "current" solution can be updated by non-LSP actions, so we need it, but we also need
+                // to merge in the changes from any mutations that have been applied to open documents
+                var solution = GetCurrentSolution();
+                solution = MergeChanges(solution, lastMutatedSolution);
```

Comment:
so we have like two mutations going on simultaneously that we're trying to reconcile?  i think i hate that.  why is it not hte case that any mutations just affect the single canonical workspace.CurrentSOlution?

Reply:
I believe this is just referencing the issue where we keep track of LSP open documents separately from the workspace's current solution - https://github.com/dotnet/roslyn/issues/45427

We should only ever have 1 LSP mutation at a time, but the workspace's current solution can change (e.g. adding project references)

Reply:
@dibarbet is correct. Solution changes can still come from the workspace, and then we'll merge in the changes for the documents that are open via LSP.